File: src/kotlin/app/src/main/kotlin/ru/renev/adventofcode2021/Day3.kt
package ru.renev.adventofcode2021

class Day3 {
    fun part1(): Int {
        println("Day 3 part 1")

        val lines = Day3Input.values.split("\n")
        val length = lines[0].length

        val gamma = CharArray(length)

        for (i in 0 until length) {
            val quantities = lines.map { it[i] }
                .groupBy { it }
                .map { it.key to it.value.count() }
                .associate { it }
            gamma[i] = if (quantities.getOrDefault('1', 0) >= quantities.getOrDefault('0', 0)) '1' else '0'
        }

        val epsilon = gamma.map { if (it == '0') '1' else '0' }.toCharArray()

        val gammaValue = String(gamma).toInt(2)
        val epsilonValue = String(epsilon).toInt(2)

        return gammaValue * epsilonValue
    }

    fun part2(): Int {
        println("Day 3 part 2")

        val lines = Day3Input.values.split("\n")
        val length = lines[0].length

        val oxygen = calculateNumber(lines, length) { count0, count1 -> if (count1 >= count0) '1' else '0' }
        val co2 = calculateNumber(lines, length) { count0, count1 -> if (count0 <= count1) '0' else '1' }

        return oxygen * co2
    }

    private fun calculateNumber(source: List<String>, length: Int, charComparison: (Int, Int) -> Char): Int {
        var currentLines = source

        for (i in 0 until length) {
            val counts = currentLines.map { it[i] }
                .groupBy { it }
                .map { it.key to it.value.count() }
                .associate { it }

            val foundChar = charComparison(counts.getOrDefault('0', 0), counts.getOrDefault('1', 0))
            currentLines = currentLines.filter { it[i] == foundChar }

            if (currentLines.count() == 1) {
                return currentLines.first().toInt(2)
            }
        }

        throw RuntimeException("Something went wrong")
    }
}

File: src/dotnet/AdventOfCode2021/AdventOfCode2021.Solutions/Day3.cs
namespace AdventOfCode2021.Solutions;

public partial class Day3 : IDay
{
    public long CalculatePartOne()
    {
        var lines = Input.Split(Environment.NewLine);
        var length = lines[0].Length;

        var gamma = new char[length];

        foreach (var i in Enumerable.Range(0, length))
        {
            var quantities = lines.Select(line => line[i])
              .GroupBy(ch => ch)
              .Select(ch => new { ch = ch.Key, count = ch.Count() })
              .ToDictionary(c => c.ch, c => c.count);

            gamma[i] = quantities['1'] >= quantities['0'] ? '1' : '0';
        }

        var epsilon = gamma.Select(ch => ch == '1' ? '0' : '1').ToArray();

        var gammaValue = Convert.ToInt32(new string(gamma), 2);
        var epsilonValue = Convert.ToInt32(new string(epsilon), 2);

        return gammaValue * epsilonValue;
    }

    public long CalculatePartTwo()
    {
        var lines = Input.Split(Environment.NewLine);
        var length = lines[0].Length;

        int CalculateNumber(string[] source, Func<(int count0, int count1), char> charComparison)
        {
            var currentLines = source;
            foreach (var i in Enumerable.Range(0, length))
            {
                var counts = currentLines.Select(line => line[i])
                  .GroupBy(ch => ch)
                  .ToDictionary(grp => grp.Key, grp => grp.Count());
                var foundChar = charComparison((counts['0'], counts['1']));
                currentLines = currentLines.Where(line => line[i] == foundChar).ToArray();

                if (currentLines.Length == 1)
                {
                    return Convert.ToInt32(currentLines[0], 2);
                }
            }

            throw new Exception("Something went wrong");
        }

        var oxygen = CalculateNumber(lines, counts => counts.count1 >= counts.count0 ? '1' : '0');
        var co2 = CalculateNumber(lines, counts => counts.count0 <= counts.count1 ? '0' : '1');

        return oxygen * co2;
    }
}
